src/lib.rs

```diff
@@ -1,14 +1,165 @@
-pub fn add(left: usize, right: usize) -> usize {
-    left + right
+use sha2::{Digest, Sha256};
+
+/// A struct representing a Merkle Tree.
+#[derive(Debug, Clone)]
+pub struct MerkleTree {
+    /// The root hash of the Merkle Tree. It is `None` if tree is empty.
+    pub root: Option<String>,
+    /// The leaf nodes of the Merkle Tree.
+    pub leaves: Vec<String>,
 }
 
-#[cfg(test)]
-mod tests {
-    use super::*;
+impl MerkleTree {
+    /// Creates a new Merkle Tree from a vector of string elements.
+    ///
+    /// # Arguments
+    ///
+    /// * `elements` - A vector of string slices representing the elements to be included in the tree.
+    ///
+    /// # Returns
+    ///
+    /// A new `MerkleTree` instance.
+    pub fn new(elements: Vec<&str>) -> Self {
+        let leaves: Vec<String> = elements.into_iter().map(Self::hash).collect();
+        let root = Self::build_tree(&leaves);
+        Self { root, leaves }
+    }
+
+    /// Builds the Merkle Tree from the given leaves.
+    ///
+    /// # Arguments
+    ///
+    /// * `leaves` - A slice of strings representing the leaf nodes.
+    ///
+    /// # Returns
+    ///
+    /// An `Option<String>` representing the root hash of the tree. It is `None` if tree is empty.
+    fn build_tree(leaves: &[String]) -> Option<String> {
+        if leaves.is_empty() {
+            return None;
+        }
+        let mut current_level = leaves.to_vec();
+        while current_level.len() > 1 {
+            let mut next_level = Vec::new();
+            for i in (0..current_level.len()).step_by(2) {
+                let left = &current_level[i];
+                let right = if i + 1 < current_level.len() {
+                    &current_level[i + 1]
+                } else {
+                    left
+                };
+                let parent_hash = Self::hash(&(left.clone() + right));
+                next_level.push(parent_hash);
+            }
+            current_level = next_level;
+        }
+        Some(current_level[0].clone())
+    }
+
+    /// Adds a new leaf element and builds the Merkle Tree again.
+    ///
+    /// # Arguments
+    ///
+    /// * `element` - A string slice representing the new element to be added.
+    pub fn add_element(&mut self, element: &str) {
+        let new_leaf = Self::hash(element);
+        self.leaves.push(new_leaf.clone());
+        self.root = Self::build_tree(&self.leaves);
+    }
+
+    /// Generates a proof that an element is contained in the Merkle Tree.
+    ///
+    /// # Arguments
+    ///
+    /// * `element` - A string slice representing the element for which the proof is to be generated.
+    ///
+    /// # Returns
+    ///
+    /// An `Option<Vec<(String, bool)>>` representing the proof, where each tuple contains a hash
+    /// and a boolean indicating whether the hash is from the left (true) or right (false) sibling.
+    pub fn generate_proof(&self, element: &str) -> Option<Vec<(String, bool)>> {
```

Comment:
I think instead of using a boolean to represent left and right siblings, you could use an `enum`. It would be more idiomatic and clearer for the library consumer.

Reply:
Done!